s3torchconnector/src/s3torchconnector/dcp/fsdp_filesystem.py

```diff
@@ -0,0 +1,246 @@
+#  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#  // SPDX-License-Identifier: BSD
+
+import io
+import logging
+import os
+from contextlib import contextmanager
+from typing import Generator, Union, Optional
+
+from tenacity import (
+    retry,
+    stop_after_attempt,
+    stop_after_delay,
+    wait_fixed,
+    wait_random,
+    retry_if_exception_type,
+)
+from torch.distributed.checkpoint.filesystem import (
+    FileSystem,
+    FileSystemReader,
+    FileSystemWriter,
+)
+
+from s3torchconnector import S3Checkpoint
+from s3torchconnector._s3client import S3Client
+from s3torchconnector._s3dataset_common import parse_s3_uri
+from s3torchconnectorclient._mountpoint_s3_client import S3Exception
+
+logger = logging.getLogger(__name__)
+
+
+class S3FileSystem(FileSystem):
+    def __init__(self, region: str, s3_client: Optional[S3Client] = None) -> None:
```

Comment:
That was already there when I picked this up; now, I realized that `S3Checkpoint` also uses an `S3Client`. Not sure it is there for testing purposes only, because we use it at many places below: in fact, I'm instead not sure as to why we have an `S3Checkpoint` to begin with, since the latter is only used for dealing with stream writer/reader, and it adds little to no value...